### PR TITLE
Multi-Currency-Checkout: Types

### PR DESCRIPTION
**Pull request by danimoh**
First PR in a series of pull requests that add the multi-currency-checkout feature.
This first PR adds the required changes for the checkout API, type definitions and parsing.
Subsequent PRs will add the corresponding UI changes.

Note that this PR is not able to build without these UI changes.
The earliest commit that is able to build is "Update Checkout to use Carousel of CheckoutOptions" in "sebastian/checkout/checkout".

As the multi-currency-checkout PRs combined form the multi-currency-checkout feature, the single PRs target a `multi-currency-checkout` feature branch which will then be merged into `master` once all single PRs have been merged together into the feature branch.
This way, merge conflicts do not need to be resolved for every single PR but only when merging the complete change set into `master` in the end.

### CODE REVIEW

File: src/lib/WalletInfoCollector.ts [hunk 1]
```diff
@@ -223,7 +223,7 @@ export default class WalletInfoCollector {
 
             const releaseKey = walletType === WalletType.BIP39
                 ? (removeKey?: boolean) => WalletInfoCollector._keyguardClient!.releaseKey(keyId!, removeKey)
-                : null;
+                : undefined;
```
Comment (sisou) [inline src/lib/WalletInfoCollector.ts]: What's the reason for this change?

File: src/lib/paymentOptions/BitcoinPaymentOptions.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        return {
+            currency: this.currency,
+            type: this.type,
+            expires: this.expires,
+            amount: this.amount.toString(),
+            protocolSpecific: {
+                fee: this.protocolSpecific.fee,
+                feePerByte: this.protocolSpecific.feePerByte,
+                recipient: this.protocolSpecific.recipient,
+            },
```
Comment (sisou) [inline src/lib/paymentOptions/BitcoinPaymentOptions.ts]: Since there is no type conversion happening, this can just be a standard assignment:

```ts
    protocolSpecific: this.protocolSpecific,
```

File: src/lib/paymentOptions/EtherPaymentOptions.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        let gasPrice: bigInt.BigInteger | undefined;
+        if (options.protocolSpecific.gasPrice !== undefined) {
+            if (!this.isNonNegativeInteger(options.protocolSpecific.gasPrice)) {
+                throw new Error('If provided, gasPrice must be a non-negative integer');
+            }
+            gasPrice = bigInt(options.protocolSpecific.gasPrice);
+        }
+
+        if (options.protocolSpecific.recipient && typeof options.protocolSpecific.recipient !== 'string') {
+            // TODO add eth address validation here?
```
Comment (sisou) [inline src/lib/paymentOptions/EtherPaymentOptions.ts]: Should be pretty straight forward to just check that it's starting with "0x" and that it's a HEX string. The checksum check (pattern of upper and lowercase letters) can propably be skipped for now.

File: src/lib/paymentOptions/EtherPaymentOptions.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    public fiatFee(fiatAmount: number): number {
+        if (!this.amount || !fiatAmount) {
+            throw new Error('amount and fiatAmount must be provided');
+        }
+
+        if (this.fee.isZero()) {
+            return 0;
+        }
```
Comment (sisou) [inline src/lib/paymentOptions/EtherPaymentOptions.ts]: Can we check for zero before checking for the `amount` and `fiatAmount`?

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    public amount: number;
+    private extraData: Uint8Array;
+
+    public constructor(options: NimiqDirectPaymentOptions, extraData: Uint8Array) {
+        super(options);
+        this.extraData = extraData;
+        this.amount = Number.parseInt(toNonScientificNumberString(options.amount), 10);
+
+        let sender: Nimiq.Address | undefined;
+        if (options.protocolSpecific.sender !== undefined) {
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]:
```suggestion
        if (options.protocolSpecific.sender) {
```

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            try {
+                sender = Nimiq.Address.fromUserFriendlyAddress(options.protocolSpecific.sender);
+            } catch (err) {
+                throw new Error('If provided, sender must be a valid user friendly address string');
+            }
+        }
+
+        let recipient: Nimiq.Address | undefined;
+        if (options.protocolSpecific.recipient !== undefined) {
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]:
```suggestion
        if (options.protocolSpecific.recipient) {
```

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            try {
+                recipient = Nimiq.Address.fromUserFriendlyAddress(options.protocolSpecific.recipient);
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: We could use `Nimiq.Address.fromString()` here to also allow HEX and base64 encodings.

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            } catch (err) {
+                throw new Error('If provided, recipient must be a valid user friendly address string');
+            }
+        }
+
+        let recipientType: Nimiq.Account.Type | undefined;
+        if (options.protocolSpecific.recipientType !== undefined) {
+            if (!Object.values(Nimiq.Account.Type).includes(options.protocolSpecific.recipientType)) {
+                throw new Error('If provided, recipientType must be a valid nimiq account type');
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]:
```suggestion
                throw new Error('If provided, recipientType must be a valid Nimiq account type');
```

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            }
+            recipientType = options.protocolSpecific.recipientType;
+        }
+
+        const flags = (options as NimiqDirectPaymentOptions).protocolSpecific.flags;
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: Type casting should be unnecessary as the argument to the constructor is already typed.

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+        let fee: number | undefined;
+        if (options.protocolSpecific.fee !== undefined) {
+            if (!this.isNonNegativeInteger(options.protocolSpecific.fee)) {
+                throw new Error('If provided, fee must be a non-negative integer');
+            }
+            fee = Number.parseInt(toNonScientificNumberString(options.protocolSpecific.fee), 10);
+        }
+
+        const requiresExtendedTransaction = extraData.length > 0
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]:
```suggestion
        const requiresExtendedTransaction = extraData.byteLength > 0
```

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            || recipientType !== undefined && recipientType !== Nimiq.Account.Type.BASIC
+            || flags !== undefined && flags !== Nimiq.Transaction.Flag.NONE;
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: Do these conditions need brackets around them? Is `&&` stronger than `||`? Are you sure?

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        // Note that the transaction size can be bigger than this, for example if the sender type the user wants to use
+        // requires an extended transaction or if an extended transaction includes a multi signature proof. The size
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: Are multisignature proofs really bigger? I thought a single-sig cannot be distinguished from a multi-sig tx?

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+                fee = Math.ceil(feePerByte * estimatedTransactionSize);
+            }
+        } else {
+            feePerByte = fee / estimatedTransactionSize;
+        }
+
+        if (options.protocolSpecific.validityDuration !== undefined
+            && !this.isNonNegativeInteger(options.protocolSpecific.validityDuration)) {
+            throw new Error('If provided, validityDuration must be a non-negative integer.');
+        }
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: I'd say a validityDuration of `0` should also not be allowed :D

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+                ),
+        };
+    }
+
+    public get currency(): Currency.NIM {
+        return Currency.NIM;
+    }
+
+    public get type(): PaymentMethod.DIRECT {
+        return PaymentMethod.DIRECT;
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: As a general question: Is this only static right now because it's the only type we support and will be made dynamic in the future, or will OASIS payments need a new PaymentOptions class?

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    public get decimals(): number {
+        return 5;
+    }
+
+    public get total(): number {
+        return this.amount + this.fee;
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: Where is `this.amount` set/validated?

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    public fiatFee(fiatAmount: number): number {
+        if (!this.amount || !fiatAmount) {
+            throw new Error('amount and fiatAmount must be provided');
+            return 0;
+        }
+        if (!this.fee) {
+            return 0;
+        }
```
Comment (sisou) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: Can `0` be returned _before_ the (fiat)amount is checked?

File: src/views/Checkout.vue [hunk 17]
```diff
@@ -268,6 +268,11 @@ export default class Checkout extends Vue {
                     validityStartHeight,
                     data: this.request.data,
                     flags: this.request.flags,
+
+                    fiatAmount: this.request.fiatAmount,
+                    fiatCurrency: this.request.fiatCurrency,
+                    time: this.request.time, // TODO convert times from server time to local time
```
Comment (sisou) [inline src/views/Checkout.vue]: `TODO`

File: src/views/SignTransactionLedger.vue [hunk 18]
```diff
@@ -85,9 +85,14 @@ import StatusScreen from '../components/StatusScreen.vue';
 import { Static } from '../lib/StaticStore';
 import { Getter } from 'vuex-class';
 import { State as RpcState } from '@nimiq/rpc';
-import { ParsedCheckoutRequest, ParsedSignTransactionRequest, RequestType } from '../lib/RequestTypes';
+import {
+    ParsedCheckoutRequest,
+    ParsedSignTransactionRequest,
+} from '../lib/RequestTypes';
```
Comment (sisou) [inline src/views/SignTransactionLedger.vue]: I would keep this in one line, even now that it got shorter.

File: src/views/SignTransactionLedger.vue [hunk 19]
```diff
@@ -148,8 +153,19 @@ export default class SignTransactionLedger extends Vue {
 
             senderUserFriendlyAddress = this.$store.state.activeUserFriendlyAddress;
 
+            /*
+             * There has to be a NIM paymentOption so find will always return a result.
+             */
+            const recipient = (this.request as ParsedSignTransactionRequest).recipient
+                || ((this.request as ParsedCheckoutRequest).paymentOptions.find(
+                        (option) => option.currency === Currency.NIM,
+                    ) as ParsedNimiqDirectPaymentOptions).protocolSpecific.recipient;
```
Comment (sisou) [inline src/views/SignTransactionLedger.vue]: I feel this finding of the Nimiq PaymentOption is so much used, it warrants its own helper function. This helper would also be typed correctly, so no need for this repetition everywhere.

File: src/views/SignTransactionLedger.vue [hunk 20]
```diff
@@ -175,17 +191,34 @@ export default class SignTransactionLedger extends Vue {
...
+        let flags: number;
+        let validityStartHeight: number;
+
         if (this.request.kind === RequestType.SIGN_TRANSACTION) {
             const signTransactionRequest = this.request as ParsedSignTransactionRequest;
+            recipient = signTransactionRequest.recipient;
+            value = signTransactionRequest.value;
+            fee = signTransactionRequest.fee;
+            flags = signTransactionRequest.flags;
             validityStartHeight = signTransactionRequest.validityStartHeight;
```
Comment (sisou) [inline src/views/SignTransactionLedger.vue]: Can be shorter:
```ts
{ recipient, value, fee, flags, validityStartHeight } = signTransactionRequest;
```

File: src/views/SignTransactionLedger.vue [hunk 21]
```diff
@@ -175,17 +191,34 @@ export default class SignTransactionLedger extends Vue {
...
         } else if (this.request.kind === RequestType.CHECKOUT) {
             const checkoutRequest = this.request as ParsedCheckoutRequest;
+            // Once currency selection is implemented likely no longer needed
+            const nimiqPaymentOption = (checkoutRequest.paymentOptions.find(
+                    (option) => option.currency === Currency.NIM,
+                ) as ParsedNimiqDirectPaymentOptions);
+            recipient = nimiqPaymentOption.protocolSpecific.recipient!;
+            value = nimiqPaymentOption.amount;
+            fee = nimiqPaymentOption.protocolSpecific.fee!;
+            flags = nimiqPaymentOption.protocolSpecific.flags;
```
Comment (sisou) [inline src/views/SignTransactionLedger.vue]: Maybe it works here, too? (I have not tried this, just feel this can work.)

```ts
{ recipient!, fee!, flags } = nimiqPaymentOption.protocolSpecific;
value = nimiqPaymentOption.amount;
```

File: src/views/SignTransactionLedger.vue [hunk 22]
```diff
@@ -175,17 +191,34 @@ export default class SignTransactionLedger extends Vue {
...
             const blockchainHeight = await network.getBlockchainHeight(); // usually instant as synced in checkout
             // The next block is the earliest for which tx are accepted by standard miners
             validityStartHeight = blockchainHeight + 1
                 - TX_VALIDITY_WINDOW
-                + checkoutRequest.validityDuration;
+                + nimiqPaymentOption.protocolSpecific.validityDuration!;
```
Comment (sisou) [inline src/views/SignTransactionLedger.vue]: The trailing `!` should not be needed, because in the ParsedNimiqPaymentOption (which this should be) this should not be optional, no?

File: tsconfig.json [hunk 23]
```diff
@@ -24,9 +24,11 @@
     },
     "lib": [
       "es2017",
+      "esnext",
```
Comment (sisou) [inline tsconfig.json]: `"esnext"` includes `"es2017"`, so `"es2017"` can be removed.

File: src/lib/paymentOptions/NimiqPaymentOptions.ts [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    public get fee(): number {
+        return this.protocolSpecific.fee;
+    }
+
+    public fiatFee(fiatAmount: number): number {
+        if (!this.amount || !fiatAmount) {
+            throw new Error('amount and fiatAmount must be provided');
+            return 0;
```
Comment (Yukiioz) [inline src/lib/paymentOptions/NimiqPaymentOptions.ts]: This `return 0;` will never be executed.

File: src/lib/RequestTypes.ts [hunk 25]
```diff
@@ -39,17 +44,92 @@ export interface ParsedSignTransactionRequest extends ParsedBasicRequest {
...
-    value: number;
-    fee: number;
+    callbackUrl?: string;
+    csrf?: string;
     data: Uint8Array;
-    flags: number;
-    validityDuration: number;
+    time: number;
+    fiatCurrency?: string;
+    fiatAmount?: number;
```
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/RequestTypes.ts]: Something along the lines of 
```ts
export type ParsedCheckoutRequest = ParsedBasicRequest & {
    data: Uint8Array;
    time: number;
    paymentOptions: AvailableParsedPaymentOptions[];
} & ({
    version: 1,
    shopLogoUrl?: string;
} | {
    version: 2,
    csrf: string;
    callbackUrl: string;
    shopLogoUrl: string;
    fiatCurrency: string;
    fiatAmount: number;
});
```
would work I think, but I also think that it works good enough as it is.

File: src/lib/RpcApi.ts [hunk 26]
```diff
@@ -225,12 +225,24 @@ export default class RpcApi {
...
                 const checkoutRequest = request as ParsedCheckoutRequest;
-                accountRequired = checkoutRequest.forceSender;
-                if (checkoutRequest.sender) {
-                    account = this._store.getters.findWalletByAddress(
-                        checkoutRequest.sender.toUserFriendlyAddress(),
-                        true,
-                    );
+                // forceSender only applies to non-multi-currency checkouts.
+                if (checkoutRequest.paymentOptions.length === 1
+                    && checkoutRequest.paymentOptions[0].currency === Currency.NIM) {
```
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/RpcApi.ts]: Also certainly does not hurt here.

File: src/lib/paymentOptions/NimiqPaymentOptions.ts (hunk 15 above)
Comment (nibhar) [reply src/lib/paymentOptions/NimiqPaymentOptions.ts]: constructor and `update()`.

File: src/lib/RequestTypes.ts [hunk 27]
```diff
@@ -39,17 +44,92 @@ export interface ParsedSignTransactionRequest extends ParsedBasicRequest {
...
+    constructor: ParsedPaymentOptionsForCurrencyAndType<C, T>;
+    new(options: PaymentOptionsForCurrencyAndType<C, T>, ...optionalArgs: any[]):
+        ParsedPaymentOptionsForCurrencyAndType<C, T>;
+    raw(): PaymentOptionsForCurrencyAndType<C, T>;
+}
+
+export abstract class ParsedPaymentOptions<C extends Currency, T extends PaymentMethod>
+    implements ParsedPaymentOptions<C, T> {
+
+    protected constructor(options: PaymentOptionsForCurrencyAndType<C, T>) {
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/RequestTypes.ts]: Because `implements ParsedPaymentOptions<C, T> {` two lines above, which is an opening curly bracket, is on the same indentation level.

File: src/lib/RequestParser.ts [hunk 28]
```diff
@@ -53,38 +64,164 @@ export class RequestParser {
...
+                        time: + new Date(),
+                        paymentOptions: [new ParsedNimiqDirectPaymentOptions({
+                            currency: Currency.NIM,
+                            type: PaymentMethod.DIRECT,
+                            amount: checkoutRequest.value.toString(),
+                            expires: 0, // unused for NimiqCheckoutRequests
+                            protocolSpecific: {
+                                recipient: checkoutRequest.recipient,
+                                recipientType: checkoutRequest.recipientType || Nimiq.Account.Type.BASIC,
+                                sender: checkoutRequest.sender,
```
[the remark this answers is not in the document]
Comment (Yukiioz) [reply src/lib/RequestParser.ts]: oh yeah you're right 👍

File: src/lib/RequestTypes.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/RequestTypes.ts]: ohhhhh... that is a continuation of the class opener... well, that is not easy to spot :D

File: src/lib/PublicRequestTypes.ts [hunk 29]
```diff
@@ -41,6 +61,85 @@ export interface CheckoutRequest extends BasicRequest {
...
+     * The csrf token, that will be transmitted for future requests to the callback url
+     */
+    csrf?: string;
+    extraData?: Uint8Array | string;
+    /**
+     * Current time in milliseconds
+     */
+    time: number;
+    /**
+     * ISO 4217 Code of the fiat currency used on the calling site.
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/PublicRequestTypes.ts]: Yes, I know, but it is meant as a helper for people who don't want to look up what ISO 4217 is.

File: src/lib/PublicRequestTypes.ts [hunk 30]
```diff
@@ -41,6 +61,85 @@ export interface CheckoutRequest extends BasicRequest {
...
+export enum PaymentMethod {
+    DIRECT,
+    OASIS,
+}
+
+export enum Currency {
+    NIM = 'nim',
+    BTC = 'btc',
+    ETH = 'eth',
+}
```
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/PublicRequestTypes.ts]: In my opinion is fine as is.

File: src/lib/RequestParser.ts [hunk 31]
```diff
@@ -53,38 +64,164 @@ export class RequestParser {
...
+                                forceSender: !!checkoutRequest.forceSender,
+                                fee: checkoutRequest.fee || 0,
+                                flags: checkoutRequest.flags || Nimiq.Transaction.Flag.NONE,
+                                validityDuration: checkoutRequest.validityDuration,
+                            },
+                        }, data)],
```
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/RequestParser.ts]: I started #351 to remove `extraData` from `MultiCurrencyCheckoutRequest`, and move it into `NimiqDirectPaymentOptions.protocolSpecific`.
As that change is not contained in this PR I opted to create it in a new branch and update it along the way while merging the remaining Checkout PR.

File: client/package.json [hunk 32]
```diff
@@ -15,7 +15,7 @@
     "rollup-plugin-node-resolve": "^5.0.0",
     "rollup-plugin-terser": "^4.0.4",
     "tslint": "^5.16.0",
-    "typescript": "^3.4.5"
+    "typescript": "^3.5.1"
```
[the remark this answers is not in the document]
Comment (danimoh) [reply client/package.json]: Didn't do so to avoid an incompatibility between new `typescript` version and `@types/node` which is a sub dependency of other dependencies.
I now updated both of them.

File: package.json [hunk 33]
```diff
@@ -15,13 +15,14 @@
     "dependencies": {
         "@nimiq/browser-warning": "^0.4.0",
         "@nimiq/iqons": "^1.5.0",
-        "@nimiq/keyguard-client": "^1.0.0",
+        "@nimiq/keyguard-client": "link:../nimiq.github.io/apps/keyguard/client",
```
[the remark this answers is not in the document]
Comment (danimoh) [reply package.json]: I'm now using a new client version.

[the remark this answers is not in the document]
Comment (sisou) [reply package.json]: Not commited yet, right?

File: src/lib/PublicRequestTypes.ts [hunk 34]
```diff
@@ -1,5 +1,24 @@
 import { WalletType } from './WalletInfo';
-import { RequestType } from './RequestTypes';
+import { NimiqProtocolSpecific, NimiqDirectPaymentOptions } from './paymentOptions/NimiqPaymentOptions';
+import { EtherProtocolSpecific, EtherDirectPaymentOptions } from './paymentOptions/EtherPaymentOptions';
+import { BitcoinProtocolSpecific, BitcoinDirectPaymentOptions } from './paymentOptions/BitcoinPaymentOptions';
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/PublicRequestTypes.ts]: I renamed them to `NimiqSpecifics`, etc.
But kept property `protocolSpecific` on `PaymentOptions` and `ParsedPaymentOptions` as it is as renaming it on the public types would be a not-backwards compatible change.

File: src/lib/PublicRequestTypes.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/PublicRequestTypes.ts]: I think it's ok the way the is. Maybe it's even good that this could be extended by fiat currencies.

File: src/lib/PublicRequestTypes.ts [hunk 35]
```diff
@@ -41,6 +61,85 @@ export interface CheckoutRequest extends BasicRequest {
...
+    NIM = 'nim',
+    BTC = 'btc',
+    ETH = 'eth',
+}
+
+export type ProtocolSpecificForCurrency<C extends Currency> =
+    C extends Currency.NIM ? NimiqProtocolSpecific
+    : C extends Currency.BTC ? BitcoinProtocolSpecific
+    : C extends Currency.ETH ? EtherProtocolSpecific
+    : {} | undefined;
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/PublicRequestTypes.ts]: Removed `{}` as valid option.

File: src/lib/PublicRequestTypes.ts [hunk 36]
```diff
@@ -41,6 +61,85 @@ export interface CheckoutRequest extends BasicRequest {
...
+     */
+    amount: string;
+    protocolSpecific: ProtocolSpecificForCurrency<C>;
+}
+
+export type AvailablePaymentOptions = NimiqDirectPaymentOptions
+                             | EtherDirectPaymentOptions
+                             | BitcoinDirectPaymentOptions;
+
+export type PaymentOptionsForCurrencyAndType<C extends Currency, T extends PaymentMethod> =
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/PublicRequestTypes.ts]: We should then consistently rename everything to *Type or *Method, including the name of the field. Either way the change is not backwards compatible.

[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/PublicRequestTypes.ts]: Renamed the enum to `PaymentType` and added a backwards compatible getter to HubApi.

File: src/lib/PublicRequestTypes.ts [hunk 37]
```diff
@@ -41,6 +61,85 @@ export interface CheckoutRequest extends BasicRequest {
...
+        : C extends Currency.ETH ? EtherDirectPaymentOptions
+        : PaymentOptions<C, T>
+    : PaymentOptions<C, T>;
+
+export interface MultiCurrencyCheckoutRequest extends BasicRequest {
+    version: 2;
+    /**
+     * Must be located on the same origin as the one the request is sent from.
+     */
+    shopLogoUrl: string;
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/lib/PublicRequestTypes.ts]: `logoUrl` wouldn't be specific enough. Maybe `recipientLogoUrl`.

[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/PublicRequestTypes.ts]: Since we already have `appName`, I think that the best descriptive variable is `appLogoUrl`. Or `appIconUrl`.

File: src/lib/paymentOptions/NimiqPaymentOptions.ts (hunk 7 above)
Comment (nibhar) [reply src/lib/paymentOptions/NimiqPaymentOptions.ts]: I do not see any occurrence of `Nimiq.Address.fromString` in the entirety of the Hub. I'd say if we want to support that we should do it in a different PR.

Comment (nibhar) [reply src/lib/paymentOptions/NimiqPaymentOptions.ts]: Since you created an issue for it I will resolve this here.

File: src/views/SignTransactionLedger.vue (hunk 20 above)
Comment (nibhar) [reply src/views/SignTransactionLedger.vue]: I think this is already done in a different PR. If not I'll take a look in the final feature branch.

File: src/lib/PublicRequestTypes.ts (hunk 37 above)
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/PublicRequestTypes.ts]: I opened #372 for this.

File: src/lib/PublicRequestTypes.ts [hunk 38]
```diff
@@ -41,6 +61,85 @@ export interface CheckoutRequest extends BasicRequest {
...
+
+export interface PaymentOptions<C extends Currency, T extends PaymentMethod> {
+    type: T;
+    currency: C;
+    expires?: number;
+    /**
+     * Amount in the smallest unit of the currency specified in `currency`.
+     * i.e Luna for Currency.NIM and Satoshi for Currency.BTC
+     */
+    amount: string;
```
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/PublicRequestTypes.ts]: Also included in #372.

File: src/lib/RequestTypes.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (nibhar) [reply src/lib/RequestTypes.ts]: To track it I also included this in #372, however I am not sure if it is worth the effort, but for sure worth investigating.

File: src/lib/paymentOptions/EtherPaymentOptions.ts (hunk 3 above)
Comment (nibhar) [reply src/lib/paymentOptions/EtherPaymentOptions.ts]: I created #373 for it, including the counterpart for BTC. I think it makes sense to if at all do this properly.
